yang/frr-zebra.yang

```diff
@@ -0,0 +1,1887 @@
+      description
+	"Multicast RPF lookup behavior.";
+    }
+
+    leaf ip-forwarding {
+      type boolean;
+      description
+	"IP forwarding status.";
+    }
+
+    leaf ipv6-forwarding {
+      type enumeration {
+	enum "unknown" {
+	  value -1;
+	  description
+	    "Unknown state.";
+	}
+	enum "off" {
+	  value 0;
+	  description
+	    "IPv6 forwarding disabled.";
+	}
+	enum "on" {
+	  value 1;
+	  description
+	    "IPv6 forwarding enabled.";
+	}
+      }
+      description
+	"IPv6 forwarding status.";
+    }
+
+    leaf workqueue-hold-timer {
+      type uint32 {
+        range "0..10000";
+      }
+      units "milliseconds";
+      default "10";
+      description
+        "Work-queue processing hold timer, in milliseconds.";
+    }
+
+    leaf zapi-packets {
+      type uint32 {
+        range "1..10000";
+      }
+      default "1000";
+      description
+        "Number of ZAPI packets to process before relinquishing
+        the main thread.";
+    }
+
+    container import-kernel-table {
+      description
+        "Parameters to use when importing IPv4 routes from a non-main kernel
+        routing table.";
+
+      leaf table-id {
+        type uint32 {
+          range "1..252";
+        }
+        description
+          "The kernel table id.";
+      }
+      leaf distance {
+        type uint32 {
+          range "1..255";
+        }
+        default "15";
+        description
+          "The admin distance to use for imported routes.";
+      }
+      leaf route-map {
+        type string;
+        description
+          "A route-map to filter imported routes.";
+      }
+    }
+
+    leaf allow-external-route-update {
+      type empty;
+      description
+        "Allow FRR-controlled routes to be overwritten by external processes";
+    }
+
+    leaf dplane-queue-limit {
+      type uint32 {
+        range "0..10000";
+      }
+      default "200";
+      description
+        "Limit on the number of updates queued to the dataplane subsystem.";
+    }
+
+    list vrf-vni-mapping {
+      description
+        "EVPN VNI mapping corresponding to a VRF.";
+      key "vrf-id";
+      leaf vrf-id {
+        type uint32;
+        description
+          "The identifier for a VRF.";
+      }
+      leaf vni-id {
+        type vni-id-type;
+        description
+          "The VNI id to map to the VRF.";
+      }
+      leaf prefix-only {
+        type empty;
+        description
+          "Prefix routes only.";
+      }
+    }
+
+    /*
+     * End of configuration attributes
+     */
+
+    /*
+     * Operational data.
+     */
+    container state {
+      config false;
+      description
+	"Operational data.";
+          
+    } // End of operational / state container
+
+  } // End of zebra container
+  
+  /*
+   * RPCs
+   */
+  rpc get-route-information {
+    description
+      "Retrieve IPv4 or IPv6 unicast routes.";
+
+    input {
+
+      choice ip-type {
+        case v4 {
+          leaf ipv4 {
```

Comment:
I haven't write/implemented RPCs yet, but it looks to me that it would be more convenient to `switch`/`case` an enumeration than check for existence of the "optional" and mutually exclusive nodes "ipv4" and "ipv6"

Reply:
Yes, I don't really know myself - part of this exercise is trying to figure out how it will work to manage the parameters for the complex show commands. I think the incoming parameters list will have only one of the two exclusive 'mandatory' nodes present - and that seems like something that's relatively easy to test for?